### ISSUE TITLE
[Benchmarks] Implement layer and data loading benchmarks

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#2730]
## Summary
Implement benchmark suite for performance regression testing.

## Components Needed
- Linear layer forward/backward benchmarks
- Activation function benchmarks
- Data loading throughput benchmarks
- Preprocessing benchmarks
- Optimizer step benchmarks

## Files Affected
- \`tests/shared/benchmarks/bench_layers.mojo\` (lines 87-89)
- \`tests/shared/benchmarks/bench_data_loading.mojo\` (lines 91-92)
- \`tests/shared/benchmarks/bench_optimizers.mojo\` (line 93)

## Related TODO Lines in todo.log
- Lines 86-93: benchmark placeholders

## Notes
All benchmarks currently return zero values (placeholders).

**Comment by mvillmow** [comment]
## Detailed Implementation Plan

Now I have a comprehensive understanding of the codebase. Let me create the detailed implementation plan:

# Implementation Plan: Layer and Data Loading Benchmarks (Issue #2730)

## Summary

This issue requires implementing real benchmark functionality for three placeholder files in `tests/shared/benchmarks/`. The existing files contain well-documented performance targets and skeletal code that currently returns zero values. All required components (Linear layer, activation functions, optimizers, data loaders) are **fully implemented** in the codebase and ready to be benchmarked.

**Key Infrastructure Available:**
- **Benchmarking Framework**: `shared/benchmarking/` provides `benchmark_function()`, `BenchmarkStatistics`, `perf_counter_ns()`
- **Linear Layer**: `shared/core/layers/linear.mojo` with `forward()` method
- **Activations**: `shared/core/activation.mojo` with 15+ activation functions (relu, sigmoid, tanh, gelu, softmax, etc.)
- **Optimizers**: `shared/training/optimizers/` with SGD, Adam, AdamW, RMSprop, LARS
- **Data Loading**: `shared/data/loaders.mojo` with `BatchLoader`, `Batch`, samplers
- **Transforms**: `shared/data/transforms.mojo` with Normalize, Resize, CenterCrop, etc.

---

## Step-by-Step Implementation Tasks

### Phase 1: Update Test Conftest Timing Utilities

**Task 1.1: Implement `measure_time` in conftest.mojo**
- **File**: `tests/shared/conftest.mojo:167-181`
- **Change**: Replace placeholder that returns 0.0 with actual timing using `perf_counter_ns()`
- **Pattern**: Follow `bench_matmul.mojo` timing pattern (warmup + measure + convert ns to ms)

**Task 1.2: Implement `measure_throughput` in conftest.mojo**
- **File**: `tests/shared/conftest.mojo:184-203`
- **Change**: Use the fixed `measure_time` function to calculate ops/second

---

### Phase 2: Implement Layer Benchmarks (`bench_layers.mojo`)

**Task 2.1: Implement `bench_linear_forward()`**
- **File**: `tests/shared/benchmarks/bench_layers.mojo:28-51`
- **Changes**:
  - Import Linear layer: `from shared.core.layers import Linear`
  - Import tensor utilities: `from shared.core import ExTensor, ones, zeros, randn`
  - Import timing: `from time import perf_counter_ns`
  - Create benchmark configurations: `[(784, 128), (3072, 512), (1024, 256)]` (in_features, out_features)
  - For each configuration:
    1. Create Linear layer
    2. Create input batch tensor `[batch_size, in_features]`
    3. Warmup (10 iterations)
    4. Measure 100 iterations
    5. Calculate throughput (samples/second = batch_size * iterations / time)
    6. Return `BenchmarkResult` with name, duration_ms, throughput

**Task 2.2: Implement `bench_linear_backward()`**
- **File**: `tests/shared/benchmarks/bench_layers.mojo:54-72`
- **Note**: Linear layer doesn't have explicit backward; benchmark `matmul` backward or skip
- **Alternative**: Benchmark gradient computation pattern using `sgd_step` with layer parameters

**Task 2.3: Implement `bench_activation_functions()`**
- **File**: `tests/shared/benchmarks/bench_layers.mojo:80-102`
- **Changes**:
  - Import activations: `from shared.core.activation import relu, relu6, sigmoid, tanh, gelu, softmax, leaky_relu`
  - Create test tensor of size 1M elements
  - Benchmark each activation:
    1. ReLU (forward + backward)
    2. Sigmoid (forward + backward)
    3. Tanh (forward + backward)
    4. GELU exact and approximate
    5. Softmax
  - Target: >10M elements/second

**Task 2.4: Add Conv2D and Pooling Benchmarks (Optional Enhancement)**
- Add benchmarks for `Conv2dLayer.forward()` from `shared/core/layers/conv2d.mojo`
- Add benchmarks for pooling operations if available

---

### Phase 3: Implement Data Loading Benchmarks (`bench_data_loading.mojo`)

**Task 3.1: Implement `bench_batch_loading_speed()`**
- **File**: `tests/shared/benchmarks/bench_data_loading.mojo:28-51`
- **Changes**:
  - Import data loading: `from shared.data.loaders import BatchLoader, Batch`
  - Import samplers: `from shared.data.samplers import SequentialSampler, RandomSampler`
  - Use `create_mock_dataloader` from conftest or create synthetic dataset
  - Benchmark configurations: `batch_size in [16, 32, 64, 128]`
  - Measure:
    1. Time to iterate through all batches
    2. Batches/second
    3. Sequential vs Random sampler overhead
  - Target: >1000 batches/second

**Task 3.2: Implement `bench_data_preprocessing()`**
- **File**: `tests/shared/benchmarks/bench_data_loading.mojo:54-72`
- **Changes**:
  - Import transforms: `from shared.data.transforms import Normalize, Resize, CenterCrop, Compose`
  - Create transform pipeline:
    ```mojo
    var normalize = Normalize(mean=0.5, std=0.5)
    var resize = Resize((224, 224))
    ```
  - Benchmark individual transforms and composed pipelines
  - Measure samples/second for each transform
  - Target: >10k samples/second

---

### Phase 4: Implement Optimizer Benchmarks (`bench_optimizers.mojo`)

**Task 4.1: Implement `bench_sgd_update_speed()`**
- **File**: `tests/shared/benchmarks/bench_optimizers.mojo:29-86`
- **Changes**:
  - Import: `from shared.training.optimizers.sgd import sgd_step, sgd_step_simple`
  - Import: `from shared.core import randn, zeros_like`
  - Benchmark configurations: `param_counts = [1_000_000, 10_000_000, 100_000_000]`
  - For each configuration:
    1. Create params tensor of size n
    2. Create gradients tensor of size n
    3. Create velocity tensor for momentum
    4. Warmup 10 iterations
    5. Measure 100 iterations of `sgd_step()`
    6. Calculate throughput (params/second)
  - Target: >1B params/second

**Task 4.2: Implement `bench_sgd_momentum_overhead()`**
- **File**: `tests/shared/benchmarks/bench_optimizers.mojo:89-128`
- **Changes**:
  - Compare `sgd_step_simple()` (no momentum) vs `sgd_step()` (with momentum=0.9)
  - Calculate overhead percentage
  - Target: <20% overhead from momentum

**Task 4.3: Implement `bench_adam_update_speed()`**
- **File**: `tests/shared/benchmarks/bench_optimizers.mojo:136-193`
- **Changes**:
  - Import: `from shared.training.optimizers.adam import adam_step`
  - Benchmark with same param counts as SGD
  - Track timestep t for bias correction
  - Target: >500M params/second

**Task 4.4: Implement `bench_adam_memory_usage()`**
- **File**: `tests/shared/benchmarks/bench_optimizers.mojo:196-239`
- **Note**: Memory profiling may not be directly available in Mojo
- **Alternative**: Calculate theoretical memory usage (3x param size for m, v, params)

**Task 4.5: Implement `bench_optimizer_comparison()`**
- **File**: `tests/shared/benchmarks/bench_optimizers.mojo:247-306`
- **Changes**:
  - Import all optimizers: SGD, Adam, AdamW, RMSprop
  - Run same benchmark for each
  - Compare relative throughput
  - Output comparison table

**Task 4.6: Implement `bench_simd_vectorization()`**
- **File**: `tests/shared/benchmarks/bench_optimizers.mojo:314-365`
- **Changes**:
  - Compare scalar vs SIMD optimizer operations
  - The optimizers already use SIMD (`multiply_simd`, `add_simd`)
  - Implement scalar baseline for comparison
  - Target: 4-8x speedup from SIMD

---

## Files to Modify/Create

### Files to Modify

| File | Lines | Changes |
|------|-------|---------|
| `tests/shared/conftest.mojo` | 167-203 | Implement `measure_time` and `measure_throughput` |
| `tests/shared/benchmarks/bench_layers.mojo` | 28-102 | Replace all placeholder functions with real implementations |
| `tests/shared/benchmarks/bench_data_loading.mojo` | 28-72 | Replace placeholder functions with real implementations |
| `tests/shared/benchmarks/bench_optimizers.mojo` | 29-365 | Replace all placeholder functions with real implementations |

### Files to Reference (Read-Only)

| File | Purpose |
|------|---------|
| `benchmarks/bench_matmul.mojo` | Reference for benchmarking patterns |
| `shared/benchmarking/__init__.mojo` | Benchmarking framework API |
| `shared/core/layers/linear.mojo` | Linear layer to benchmark |
| `shared/core/activation.mojo` | Activation functions to benchmark |
| `shared/training/optimizers/sgd.mojo` | SGD optimizer to benchmark |
| `shared/training/optimizers/adam.mojo` | Adam optimizer to benchmark |
| `shared/data/loaders.mojo` | Data loader to benchmark |
| `shared/data/transforms.mojo` | Transforms to benchmark |

---

## Testing Approach

### Unit Testing

1. **Verify Each Benchmark Produces Non-Zero Results**
   - Run each benchmark file independently
   - Assert results have `duration_ms > 0.0` and `throughput > 0.0`

2. **Verify Correctness Before Benchmarking**
   - Follow `bench_matmul.mojo` pattern: verify operations produce correct results before timing
   - For Linear: verify output shape is correct
   - For activations: verify known input/output pairs

### Integration Testing

1. **Run All Benchmarks End-to-End**
   ```bash
   pixi run mojo run tests/shared/benchmarks/bench_layers.mojo
   pixi run mojo run tests/shared/benchmarks/bench_data_loading.mojo
   pixi run mojo run tests/shared/benchmarks/bench_optimizers.mojo
   ```

2. **Verify Performance Targets**
   - Linear forward: >1M samples/second
   - Activation functions: >10M elements/second
   - Batch loading: >1000 batches/second
   - SGD: >1B params/second
   - Adam: >500M params/second

### Regression Testing

1. **CI Integration**
   - Add benchmark runs to CI (optional, not blocking)
   - Store baseline results for regression detection

---

## Success Criteria

### Functional Requirements

- [ ] All three benchmark files (`bench_layers.mojo`, `bench_data_loading.mojo`, `bench_optimizers.mojo`) execute without errors
- [ ] All placeholder functions replaced with real implementations
- [ ] All benchmarks return non-zero `BenchmarkResult` values
- [ ] `conftest.mojo` timing utilities (`measure_time`, `measure_throughput`) work correctly

### Performance Targets

- [ ] Linear layer forward: >1M samples/second on test hardware
- [ ] Backward pass: within 2x of forward pass time
- [ ] Activation functions: >10M elements/second
- [ ] Batch loading: >1000 batches/second
- [ ] Data preprocessing: >10k samples/second
- [ ] SGD: >1B parameters/second
- [ ] Adam: >500M parameters/second
- [ ] SIMD speedup: 4-8x over scalar operations

### Code Quality

- [ ] Follows existing `bench_matmul.mojo` patterns for consistency
- [ ] Includes warmup iterations before measurement
- [ ] Uses `perf_counter_ns()` for high-resolution timing
- [ ] Proper error handling for tensor operations
- [ ] Clear output format showing benchmark results

---

## Resource Usage

| Resource | Estimate |
|----------|----------|
| Files Modified | 4 |
| Lines Added/Changed | ~500-700 |
| Tool Calls Used | 18 |
| Tool Calls Remaining | 32 |
| Estimated Implementation Time | 2-3 hours |
